File: scripts/inventory_frontmatter.py
#!/usr/bin/env python3
"""
Inventory frontmatter keys across all agent artefacts.

Reads .agent-src.uncompressed/{skills,rules,commands,personas}, parses the
YAML frontmatter of every file, and prints per-type:

- total file count
- every key observed, with count and percentage
- sample values (up to 3) per key

Output is Markdown on stdout, intended to be captured into
`agents/docs/frontmatter-contract.md` as raw material for Phase 1 of the
frontmatter-schema roadmap.

Stdlib-only. No PyYAML — we do a simple line-based parse sufficient for
our frontmatter shapes (flat keys, inline lists, block lists, one nested
`execution:` block).
"""

from __future__ import annotations

import re
import sys
from collections import Counter, defaultdict
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
SRC = ROOT / ".agent-src.uncompressed"

FRONTMATTER_RE = re.compile(r"^---\n(.*?)\n---\n", re.DOTALL)


def extract_frontmatter(text: str) -> str | None:
    match = FRONTMATTER_RE.search(text)
    return match.group(1) if match else None


def parse_frontmatter_keys(fm: str) -> dict[str, str]:
    """Return a flat {key: raw_value_string} for a frontmatter block.

    For nested blocks (e.g. `execution:`), the nested keys are flattened
    with a dot notation: `execution.type`, `execution.handler`, etc.
    Inline lists (`personas: [a, b]`) and block lists (`- a\n- b`) are
    rendered as their raw value.
    """
    result: dict[str, str] = {}
    lines = fm.splitlines()
    i = 0
    current_nested: str | None = None
    current_list_key: str | None = None
    list_buffer: list[str] = []

    def flush_list() -> None:
        nonlocal current_list_key, list_buffer
        if current_list_key is not None:
            result[current_list_key] = "[" + ", ".join(list_buffer) + "]"
        current_list_key = None
        list_buffer = []

    while i < len(lines):
        line = lines[i]
        stripped = line.strip()

        if not stripped or stripped.startswith("#"):
            i += 1
            continue

        # Top-level key (no leading whitespace)
        if line and not line[0].isspace():
            flush_list()
            current_nested = None
            m = re.match(r"^([\w-]+):\s*(.*?)\s*$", line)
            if m:
                key, value = m.group(1), m.group(2)
                if value == "" or value == "|":
                    # Could start a nested block OR a list. Look ahead.
                    nxt = lines[i + 1].strip() if i + 1 < len(lines) else ""
                    if nxt.startswith("- "):
                        current_list_key = key
                    else:
                        current_nested = key
                        result[key] = "{nested}"
                else:
                    result[key] = value
        # Nested (indented) key
        elif current_nested is not None:
            m = re.match(r"^\s+([\w-]+):\s*(.*?)\s*$", line)
            if m:
                key, value = m.group(1), m.group(2)
                result[f"{current_nested}.{key}"] = value or "{nested}"
        # Block list item
        elif current_list_key is not None and stripped.startswith("- "):
            list_buffer.append(stripped[2:].strip())

        i += 1

    flush_list()
    return result


def gather_files(artefact_dir: Path, pattern: str) -> list[Path]:
    if not artefact_dir.exists():
        return []
    files = [f for f in artefact_dir.rglob(pattern) if not f.is_symlink()]
    return sorted(files)


def inventory_type(name: str, files: list[Path]) -> None:
    total = len(files)
    print(f"### {name} — {total} files\n")
    if total == 0:
        print("_(no files)_\n")
        return

    key_counts: Counter[str] = Counter()
    key_value_counts: dict[str, Counter[str]] = defaultdict(Counter)

    for f in files:
        text = f.read_text(encoding="utf-8")
        fm = extract_frontmatter(text)
        if fm is None:
            continue
        parsed = parse_frontmatter_keys(fm)
        for k, v in parsed.items():
            key_counts[k] += 1
            # Record distinct values for enum detection; truncate to keep
            # the table readable and strip surrounding quotes.
            normalized = v.strip('"').strip("'")[:80] if v else "{empty}"
            key_value_counts[k][normalized] += 1

    print("| key | count | % | status | distinct values (count) |")
    print("|---|---:|---:|---|---|")
    for key, count in sorted(key_counts.items(), key=lambda kv: (-kv[1], kv[0])):
        pct = count * 100 // total
        status = "required" if pct >= 95 else "optional"
        values = key_value_counts[key]
        if len(values) <= 8:
            rendered = " · ".join(f"`{val}` ({n})" for val, n in values.most_common())
        else:
            top = values.most_common(5)
            rendered = " · ".join(f"`{val}` ({n})" for val, n in top)
            rendered += f" · … +{len(values) - 5} more"
        print(f"| `{key}` | {count} | {pct}% | {status} | {rendered} |")
    print()


def main() -> int:
    print("# Frontmatter inventory (generated)\n")
    print("Generated by `scripts/inventory_frontmatter.py`. Raw material for")
    print("Phase 1 of the frontmatter-schema roadmap. Do not edit by hand.\n")

    inventory_type("skills", gather_files(SRC / "skills", "SKILL.md"))
    inventory_type("rules", gather_files(SRC / "rules", "*.md"))
    inventory_type("commands", gather_files(SRC / "commands", "*.md"))
    inventory_type("personas", [
        f for f in gather_files(SRC / "personas", "*.md")
        if f.name.lower() != "readme.md"
    ])
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/scripts/inventory_frontmatter.ts
#!/usr/bin/env node
/**
 * Inventory frontmatter keys across all agent artefacts.
 *
 * TypeScript twin of `src/scripts/inventory_frontmatter.py` (ADR-096 —
 * Python→TS migration, Phase 8 / Wave 8c). Mirrors the CLI contract
 * EXACTLY: no flags, exit code 0, byte-identical Markdown stdout. No
 * behaviour changes — latent bugs (line-based frontmatter parse, the
 * `value or "{empty}"` falsiness on `"0"` / `""`, integer `//` percent)
 * replicated.
 *
 * Reads `.agent-src.uncondensed/{skills,rules,commands,personas}`, parses
 * the YAML frontmatter of every file, and prints per-type:
 *
 * - total file count
 * - every key observed, with count and percentage
 * - sample values (up to 3) per key
 *
 * Output is Markdown on stdout, intended to be captured into
 * `agents/reference/docs/frontmatter-contract.md` as raw material for
 * Phase 1 of the frontmatter-schema roadmap.
 *
 * Stdlib-only (no YAML library) — a simple line-based parse sufficient for
 * our frontmatter shapes (flat keys, inline lists, block lists, one nested
 * `execution:` block).
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const _HERE = fileURLToPath(import.meta.url);
// src/scripts/inventory_frontmatter.ts → parents[2] is the repo root.
const ROOT = path.resolve(path.dirname(_HERE), '..', '..');
const SRC = path.join(ROOT, '.agent-src.uncondensed');

// Python: re.compile(r"^---\n(.*?)\n---\n", re.DOTALL)
const FRONTMATTER_RE = /^---\n([\s\S]*?)\n---\n/;

function extract_frontmatter(text: string): string | null {
    const m = FRONTMATTER_RE.exec(text);
    return m ? m[1]! : null;
}

const _TOPLEVEL_KEY_RE = /^([\w-]+):\s*(.*?)\s*$/;
const _NESTED_KEY_RE = /^\s+([\w-]+):\s*(.*?)\s*$/;

/**
 * Return a flat {key: raw_value_string} for a frontmatter block.
 *
 * For nested blocks (e.g. `execution:`), the nested keys are flattened with
 * dot notation: `execution.type`, `execution.handler`, etc. Inline lists
 * (`personas: [a, b]`) and block lists (`- a\n- b`) are rendered as their
 * raw value.
 */
function parse_frontmatter_keys(fm: string): Map<string, string> {
    const result = new Map<string, string>();
    const lines = fm.split('\n');
    let i = 0;
    let current_nested: string | null = null;
    let current_list_key: string | null = null;
    let list_buffer: string[] = [];

    const flush_list = (): void => {
        if (current_list_key !== null) {
            result.set(current_list_key, '[' + list_buffer.join(', ') + ']');
        }
        current_list_key = null;
        list_buffer = [];
    };

    while (i < lines.length) {
        const line = lines[i]!;
        const stripped = _pyStrip(line);

        if (!stripped || stripped.startsWith('#')) {
            i += 1;
            continue;
        }

        // Top-level key (no leading whitespace). Python: `line[0].isspace()`.
        if (line.length > 0 && !_isPySpace(line[0]!)) {
            flush_list();
            current_nested = null;
            const m = _TOPLEVEL_KEY_RE.exec(line);
            if (m) {
                const key = m[1]!;
                const value = m[2]!;
                if (value === '' || value === '|') {
                    // Could start a nested block OR a list. Look ahead.
                    const nxt = i + 1 < lines.length ? _pyStrip(lines[i + 1]!) : '';
                    if (nxt.startsWith('- ')) {
                        current_list_key = key;
                    } else {
                        current_nested = key;
                        result.set(key, '{nested}');
                    }
                } else {
                    result.set(key, value);
                }
            }
        } else if (current_nested !== null) {
            // Nested (indented) key.
            const m = _NESTED_KEY_RE.exec(line);
            if (m) {
                const key = m[1]!;
                const value = m[2]!;
                result.set(`${current_nested}.${key}`, value || '{nested}');
            }
        } else if (current_list_key !== null && stripped.startsWith('- ')) {
            // Block list item.
            list_buffer.push(_pyStrip(stripped.slice(2)));
        }

        i += 1;
    }

    flush_list();
    return result;
}

function gather_files(artefact_dir: string, predicate: (basename: string) => boolean): string[] {
    if (!fs.existsSync(artefact_dir)) {
        return [];
    }
    const files: string[] = [];
    const walk = (dir: string): void => {
        let entries: fs.Dirent[];
        try {
            entries = fs.readdirSync(dir, { withFileTypes: true });
        } catch {
            return;
        }
        for (const e of entries) {
            const full = path.join(dir, e.name);
            // Python: `if not f.is_symlink()` — drop symlinks (dirs + files).
            if (e.isSymbolicLink()) {
                continue;
            }
            if (e.isDirectory()) {
                walk(full);
            } else if (e.isFile() && predicate(e.name)) {
                files.push(full);
            }
        }
    };
    walk(artefact_dir);
    return _pySortPaths(files);
}

/**
 * pathlib `sorted(list[Path])` — component-wise comparison of the path
 * parts. Mirrors how Python orders Path objects.
 */
function _pySortPaths(paths: string[]): string[] {
    return [...paths].sort((a, b) => {
        const pa = a.split(path.sep);
        const pb = b.split(path.sep);
        const n = Math.min(pa.length, pb.length);
        for (let i = 0; i < n; i++) {
            const c = _pyStrCmp(pa[i]!, pb[i]!);
            if (c !== 0) {
                return c;
            }
        }
        return pa.length - pb.length;
    });
}

function _pyStrCmp(a: string, b: string): number {
    const ca = Array.from(a);
    const cb = Array.from(b);
    const n = Math.min(ca.length, cb.length);
    for (let i = 0; i < n; i++) {
        const x = ca[i]!.codePointAt(0)!;
        const y = cb[i]!.codePointAt(0)!;
        if (x !== y) {
            return x - y;
        }
    }
    return ca.length - cb.length;
}

const _OUT: string[] = [];
function emit(line = ''): void {
    _OUT.push(line);
}

function inventory_type(name: string, files: string[]): void {
    const total = files.length;
    emit(`### ${name} — ${total} files\n`);
    if (total === 0) {
        emit('_(no files)_\n');
        return;
    }

    // Counter[str] — insertion-ordered key counts.
    const key_counts = new Map<string, number>();
    // dict[str, Counter[str]] — per-key value counts, insertion-ordered.
    const key_value_counts = new Map<string, Map<string, number>>();

    for (const f of files) {
        const text = fs.readFileSync(f, 'utf-8');
        const fm = extract_frontmatter(text);
        if (fm === null) {
            continue;
        }
        const parsed = parse_frontmatter_keys(fm);
        for (const [k, v] of parsed) {
            key_counts.set(k, (key_counts.get(k) ?? 0) + 1);
            // Record distinct values for enum detection; truncate to keep
            // the table readable and strip surrounding quotes.
            // Python: `v.strip('"').strip("'")[:80] if v else "{empty}"`.
            const normalized = v ? _sliceCodepoints(_pyStripChars(_pyStripChars(v, '"'), "'"), 80) : '{empty}';
            const inner = key_value_counts.get(k) ?? new Map<string, number>();
            inner.set(normalized, (inner.get(normalized) ?? 0) + 1);
            key_value_counts.set(k, inner);
        }
    }

    emit('| key | count | % | status | distinct values (count) |');
    emit('|---|---:|---:|---|---|');
    // sorted(key_counts.items(), key=lambda kv: (-kv[1], kv[0])).
    const sortedKeys = [...key_counts.entries()].sort((a, b) => {
        if (a[1] !== b[1]) {
            return b[1] - a[1]; // -count ascending == count descending
        }
        return _pyStrCmp(a[0], b[0]);
    });
    for (const [key, count] of sortedKeys) {
        // Python integer division: count * 100 // total.
        const pct = Math.floor((count * 100) / total);
        const status = pct >= 95 ? 'required' : 'optional';
        const values = key_value_counts.get(key)!;
        let rendered: string;
        if (values.size <= 8) {
            rendered = _mostCommon(values)
                .map(([val, n]) => `\`${val}\` (${n})`)
                .join(' · ');
        } else {
            const top = _mostCommon(values).slice(0, 5);
            rendered = top.map(([val, n]) => `\`${val}\` (${n})`).join(' · ');
            rendered += ` · … +${values.size - 5} more`;
        }
        emit(`| \`${key}\` | ${count} | ${pct}% | ${status} | ${rendered} |`);
    }
    emit();
}

/**
 * Mirror collections.Counter.most_common(): sort by count descending,
 * ties keep insertion order (Python's Counter.most_common is stable on
 * insertion order for equal counts via heapq/sorted stability).
 */
function _mostCommon(counter: Map<string, number>): [string, number][] {
    const indexed = [...counter.entries()].map((e, idx) => ({ e, idx }));
    indexed.sort((a, b) => {
        if (a.e[1] !== b.e[1]) {
            return b.e[1] - a.e[1];
        }
        return a.idx - b.idx;
    });
    return indexed.map((x) => x.e);
}

export function main(): number {
    _OUT.length = 0;
    emit('# Frontmatter inventory (generated)\n');
    emit('Generated by `scripts/inventory_frontmatter.py`. Raw material for');
    emit('Phase 1 of the frontmatter-schema roadmap. Do not edit by hand.\n');

    inventory_type('skills', gather_files(path.join(SRC, 'skills'), (b) => b === 'SKILL.md'));
    inventory_type('rules', gather_files(path.join(SRC, 'rules'), (b) => b.endsWith('.md')));
    inventory_type('commands', gather_files(path.join(SRC, 'commands'), (b) => b.endsWith('.md')));
    inventory_type(
        'personas',
        gather_files(path.join(SRC, 'personas'), (b) => b.endsWith('.md')).filter(
            (f) => path.basename(f).toLowerCase() !== 'readme.md',
        ),
    );

    process.stdout.write(_OUT.join('\n') + '\n');
    return 0;
}

// --- Python string-method helpers --------------------------------------------

/** Mirror Python str.strip() default — strip leading + trailing whitespace. */
function _pyStrip(s: string): string {
    return s.replace(/^\s+/u, '').replace(/\s+$/u, '');
}

/** Mirror Python str.strip(chars) — strip any of the given chars from both ends. */
function _pyStripChars(s: string, chars: string): string {
    const set = new Set(Array.from(chars));
    const arr = Array.from(s);
    let start = 0;
    let end = arr.length;
    while (start < end && set.has(arr[start]!)) {
        start += 1;
    }
    while (end > start && set.has(arr[end - 1]!)) {
        end -= 1;
    }
    return arr.slice(start, end).join('');
}

/** Mirror Python str[:n] on code points. */
function _sliceCodepoints(s: string, n: number): string {
    return Array.from(s).slice(0, n).join('');
}

/** Mirror Python str.isspace() for a single char (subset: common whitespace). */
function _isPySpace(ch: string): boolean {
    return /\s/u.test(ch);
}

if (import.meta.url === pathToFileURL(process.argv[1] ?? '').href) {
    process.exitCode = main();
}
